### ISSUE TITLE
Error: the API version in the data (tekton.dev/v1beta1) does not match the expected API version (tekton.dev/v1)

### ISSUE THREAD

**Issue by akumacxd** [tektoncd/cli#1837]
# Versions and Operating System

- Kubernetes version:

  **Output of `kubectl version`:**

```
# kubectl get nodes
NAME        STATUS   ROLES               AGE   VERSION
rancher02   Ready    controlplane,etcd   31d   v1.24.6
rancher03   Ready    worker              31d   v1.24.6
rancher04   Ready    worker              31d   v1.24.6
rancher05   Ready    worker              31d   v1.24.6
```

- Tekton Pipeline version:

```
# tkn version
Client version: 0.28.0
Pipeline version: v0.43.0
Dashboard version: v0.31.0
```

- Operating System:

```
# cat /etc/os-release 
NAME="SLES"
VERSION="15-SP4"
VERSION_ID="15.4"
PRETTY_NAME="SUSE Linux Enterprise Server 15 SP4"
ID="sles"
ID_LIKE="suse"
ANSI_COLOR="0;32"
CPE_NAME="cpe:/o:suse:sles:15:sp4"
DOCUMENTATION_URL="https://documentation.suse.com/"
```

# Error Info

```
# tkn task start --showlog workspace-demo -p target="SUSE" -w name=messages,emptyDir=""
Error: the API version in the data (tekton.dev/v1beta1) does not match the expected API version (tekton.dev/v1)

# kubectl api-versions | grep tekton
resolution.tekton.dev/v1alpha1
resolution.tekton.dev/v1beta1
tekton.dev/v1
tekton.dev/v1alpha1
tekton.dev/v1beta1

# kubectl api-resources --api-group=tekton.dev
NAME                   SHORTNAMES   APIVERSION            NAMESPACED   KIND
clustertasks                        tekton.dev/v1beta1    false        ClusterTask
customruns                          tekton.dev/v1beta1    true         CustomRun
pipelineresources                   tekton.dev/v1alpha1   true         PipelineResource
pipelineruns           pr,prs       tekton.dev/v1         true         PipelineRun
pipelines                           tekton.dev/v1         true         Pipeline
runs                                tekton.dev/v1alpha1   true         Run
taskruns               tr,trs       tekton.dev/v1         true         TaskRun
tasks                               tekton.dev/v1         true         Task
verificationpolicies                tekton.dev/v1alpha1   true         VerificationPolicy
```

## Pipeline
```
apiVersion: tekton.dev/v1
kind: Task
metadata:
  name: workspace-demo
spec:
  params:
  - name: target
    type: string
    default: MageEdu
  steps:
    - name: write-message
      image: alpine:3.16
      script: |
        #!/bin/sh
        set -xe
        if [ "$(workspaces.messages.bound)" == "true" ] ; then
          echo "Hello $(params.target)" > $(workspaces.messages.path)/message
          cat $(workspaces.messages.path)/message
        fi
        echo "Mount Path: $(workspaces.messages.path)"
        echo "Volume Name: $(workspaces.messages.volume)"
  workspaces:
    - name: messages
      description: |
        The folder where we write the message to. If no workspace
        is provided then the message will not be written.
      optional: true
      mountPath: /data
```

**Comment by vdemeester** [comment]
@akumacxd thanks for the report. We didn't yet update the `cli` to the latest release of `tektoncd/pipeline` dependency. That said, it's a bit weird as the cli should still query `v1beta1` 🤔

**Comment by chengjoey** [comment]
> @akumacxd thanks for the report. We didn't yet update the `cli` to the latest release of `tektoncd/pipeline` dependency. That said, it's a bit weird as the cli should still query `v1beta1` 🤔

![image](https://user-images.githubusercontent.com/30427474/210329289-b9c4b189-22f0-4bab-8308-073d5dbeb98c.png)
https://github.com/tektoncd/hub/blob/0762ee16dea1e3718e7b45559e1661ea4d925f5b/api/pkg/cli/installer/gvr.go#L30-L46
hi @vdemeester  after upgrading the tekton-pipeline to v0.43.0, the queried gvr version is already v1, so using v1beta1 to create a v1 task failed

**Comment by bendory** [comment]
IIUC, the root cause here is that the CLI needs to be updated ~concurrently with the latest pipelines API release? If that's the case, this sounds very much like a dependency we want to decouple -- I would certainly not expect our users to be able to rev all their pipeline installations concurrently with all their CLI installations.

**Comment by vdemeester** [comment]
@bendory same. Ideally the current cli should be able to target any pipeline version that still serve `v1beta1`. We may have some things to fix in our cli, because for example, doing `tkn task start` with the latest `tkn` release should create `v1beta1` object as.. `tkn` doesn't know `v1` yet.

cc @piyush-garg

**Comment by lbernick** [comment]
I think the issue in the screenshot @chengjoey provided is that the hub installer package [uses the discovery client to determine what api version to use in the create request](https://github.com/tektoncd/hub/blob/0762ee16dea1e3718e7b45559e1661ea4d925f5b/api/pkg/cli/installer/kube_action.go#L31), but still includes the full content including apiversion of the file from the catalog. So the hub is issuing a create request for a v1 Task, but the unstructured bytes of the create request contain version v1beta1, and the api server rejects the request. Seems like it could be fixed by getting the api version from the data rather than using the discovery client?

**Comment by Yongxuanzhang** [comment]
I think @chengjoey's [fix](https://github.com/tektoncd/hub/pull/746) in hub should also work for cli repo too.

**Comment by wlynch** [comment]
Ran into this trying to release chains. 😢

I can probably work around this using regular kubectl, but any update on a fix for tkn itself?

**Comment by vdemeester** [comment]
cc @piyush-garg 
It's being fixed, but not released yet 😅

**Comment by drriguz** [comment]
Aha, I got the same error here, hope the fix could be released soon!

**Comment by afrittoli** [comment]
@vdemeester @piyush-garg Are we going to backport the fix to older `tkn` releases, or is the upcoming `tkn` release going to be a hard requirement to use pipelines v0.43+ ?

**Comment by vdemeester** [comment]
We need to look into how hard it is to backport it, but if we can we'll probably only backport to v0.27 (LTS).

**Comment by 524c** [comment]
I have a similar problem, in the first steps, tests here.

```
❯ tkn hub install task kubernetes-actions
Error: the API version in the data (tekton.dev/v1beta1) does not match the expected API version (tekton.dev/v1)

```
```
❯ tkn version
Client version: 0.29.1
Pipeline version: v0.45.0
Triggers version: v0.22.2
Dashboard version: v0.33.0
```

**Comment by 524c** [comment]
> I have a similar problem, in the first steps, tests here.
> 
> ```
> ❯ tkn hub install task kubernetes-actions
> Error: the API version in the data (tekton.dev/v1beta1) does not match the expected API version (tekton.dev/v1)
> ```
> 
> ```
> ❯ tkn version
> Client version: 0.29.1
> Pipeline version: v0.45.0
> Triggers version: v0.22.2
> Dashboard version: v0.33.0
> ```
> 
> ```
> ❯ uname -a
> Darwin mac2.test 22.1.0 Darwin Kernel Version 22.1.0: Sun Oct  9 20:14:54 PDT 2022; root:xnu-8792.41.9~2/RELEASE_X86_64 x86_64
> ```

To get around this was easy, I ran the manifests directly.

**Comment by mouuii** [comment]
when to fix this bug , where's the pr link ?

**Comment by drriguz** [comment]
> Ran into this while releasing pipelines `0.46.0` 😞

Any updates from the cli team? If it's not ready to release a fixed version of CLI, can we build from source as a workaround?

**Comment by piyush-garg** [comment]
Fixed in 

 - https://github.com/tektoncd/cli/pull/1953
 - https://github.com/tektoncd/cli/pull/1954
 - https://github.com/tektoncd/cli/pull/1958
 
 Available in v0.30.0 release
